### PR TITLE
Support [patch] in .cargo/config files

### LINKED ISSUE #5539: Allow `[patch]` (and `[replace]`) sections in Cargo config.
Scenario:
 * You have several smaller libraries you are developing together.
 * You don't want to, or can't, use a workspace.
 * You don't want to have to keep changing and `git stash`ing `[patch]` sections in the scattered `Cargo.toml` files (you are bound to commit them accidentally at some point).

Wouldn't it be super useful if we could put `[patch]` (and maybe `[replace]` too) sections into our `~/.cargo/config`? I'd certainly use that.

Thanks

### PR DESCRIPTION
**Pull request by jonhoo**
This patch adds support for `[patch]` sections in `.cargo/config.toml`
files. Patches from config files defer to `[patch]` in `Cargo.toml` if
both provide a patch for the same crate.

The current implementation merge config patches into the workspace
manifest patches. It's unclear if that's the right long-term plan, or
whether these patches should be stored separately (though likely still
in the manifest). Regardless, they _should_ likely continue to be
parsed when the manifest is parsed so that errors and such occur in the
same place regardless of where a patch is specified.

Fixes #5539.

### CODE REVIEW

**Comment by alexcrichton** [comment]
I'm a bit wary about modifying the patch tables so deep in the manifest, it seems like this would emulate a `[patch]` in all crates.io dependencies decoded as well I think? Would it be possible to refactor so that when the `[patch]` entries are registered with the registry that's where the config-based patches are registered?

**Comment by jonhoo** [comment]
Good point! I placed the parsing there because I wanted to re-use `to_dependency`, which takes a `Context`, which is only constructed in that scope. I've re-organized slightly in 140a7707adbfb2085e8d24e5900d79bfe49cdd6d so that the parsing still happens in the same place (where we have `Config`), but the actual merging of dependencies only happens on resolve. What do you think?

I'm also a little unsure about
https://github.com/rust-lang/cargo/blob/140a7707adbfb2085e8d24e5900d79bfe49cdd6d/src/cargo/core/workspace.rs#L394-L398

The patch tables are stored as `BTreeMap<String, BTreeMap<String, TomlDependency>>` which implies that each key in the patch table has to be unique, but I'm concerned about what happens if someone tries to patch, say, two versions of the same package using different patch names:
```toml
# Cargo.toml
[patch.crates-io]
nom5 = { ..., version = "5" }
nom6 = { ..., version = "6" }
```
```toml
# .cargo/config
[patch.crates-io]
nom = { ..., version = "5" }
```
Can that even happen? What's the expected outcome in this case?

File: src/cargo/core/workspace.rs [hunk 1]
```diff
@@ -365,11 +366,46 @@ impl<'cfg> Workspace<'cfg> {
     /// Returns the root `[patch]` section of this workspace.
     ///
     /// This may be from a virtual crate or an actual crate.
-    pub fn root_patch(&self) -> &HashMap<Url, Vec<Dependency>> {
-        match self.root_maybe() {
+    pub fn root_patch(&self) -> Cow<'_, HashMap<Url, Vec<Dependency>>> {
```
Comment (alexcrichton) [inline src/cargo/core/workspace.rs:369]: This is called so rarely it's fine to skip the `Cow` here

File: src/cargo/util/toml/mod.rs [hunk 2]
```diff
@@ -1553,6 +1556,23 @@ impl TomlManifest {
         Ok(patch)
     }
 
+    fn patch(&self, cx: &mut Context<'_, '_>) -> CargoResult<HashMap<Url, Vec<Dependency>>> {
+        let _ = cx.config.maybe_init_patch(|| {
```
Comment (alexcrichton) [inline src/cargo/util/toml/mod.rs:1560]: To me at least this feels somewhat backwards in terms of how initialization/parsing of these sorts of config tables happens. This is just sort of randomly buried in unrelated code and has a very "bolted-on" feeling.

Could this be refactored so the `Config` field could be parsed on-demand instead of sort of off-on-the-side like this?

Comment (jonhoo) [reply src/cargo/util/toml/mod.rs:1560]: I agree, it does feel a little bolted-on. The reason it is this way is because of `DetailedTomlDependency::to_dependency`, which takes a `Context`, which is only ever constructed "down here". I suppose we could construct a `Context` later on-demand, though that would mean the on-demand path _also_ needs to:

Populate the platform

https://github.com/rust-lang/cargo/blob/ce0b8e8574796a951ebe7ed872f1e2f8cc538901/src/cargo/util/toml/mod.rs#L1227-L1228

Deal with `to_dependency` changing `Context::nested_paths`:

https://github.com/rust-lang/cargo/blob/ce0b8e8574796a951ebe7ed872f1e2f8cc538901/src/cargo/util/toml/mod.rs#L1785

And extract warnings and errors and push them as appropriate:

https://github.com/rust-lang/cargo/blob/ce0b8e8574796a951ebe7ed872f1e2f8cc538901/src/cargo/util/toml/mod.rs#L1371-L1376

That's not insurmountable, but it would be some code repetition. It would also mean that warnings/errors in `.cargo/config`'s `[patch]` would be handled at a different "time" in the process than those in `Cargo.toml`, which may in turn mean that a patch in one or the other manifest very differently. Not sure how much of a problem that is?

The other way to look at this is that this is the _right_ place for initialization to happen: all patch sections, regardless of where they come from, get processed at the same time. We just need to store the config patches somewhere separate so that we can get them back later. It only really looks weird because of the fact that we store it in `Config`. If we instead store the config patch in the `Manifest` (alongside but not merged with) the regular `[patch]`, it'd be less of a problem. The downside with _that_ is that we'd then have to modify more code paths to ensure that they carry along this extra manifest information all the way to where we apply the patches. But maybe that's a nicer approach?

Comment (alexcrichton) [reply src/cargo/util/toml/mod.rs:1560]: Yeah I understand the wonkiness today, and in some sense this wonkiness is beacuse this feature is going against how Cargo is designed (and hence how the code is structured). I don't really have a preference on how to organize this but I would basically just prefer that refactoring was done (if necessary), to give this less of a bolted on feeling.

Comment (jonhoo) [reply src/cargo/util/toml/mod.rs]: I have a different approach a go in https://github.com/rust-lang/cargo/pull/9204/commits/6995e1dd1b89d919aa0e2f8443c7cbaf4931bb32 that I think you'll like better. Let me know what you think!

File: src/cargo/core/workspace.rs [hunk 3]
```diff
@@ -362,14 +362,107 @@ impl<'cfg> Workspace<'cfg> {
...
+                            source,
+                            &mut nested_paths,
+                            self.config,
+                            &mut warnings,
+                            /* platform */ None,
+                            // NOTE: Since we use ConfigRelativePath, this root isn't used as
+                            // any relative paths are resolved before they'd be joined with root.
+                            self.root(),
+                            self.unstable_features(),
+                            None,
```
Comment (alexcrichton) [inline src/cargo/core/workspace.rs:408]: Could this have a small comment indicating what it's `None` for?

File: src/cargo/core/workspace.rs [hunk 4]
```diff
@@ -362,14 +362,107 @@ impl<'cfg> Workspace<'cfg> {
...
+            );
+        }
+
+        for message in warnings {
+            self.config
+                .shell()
+                .warn(format!("[patch] in cargo config: {}", message))?
+        }
+
+        let _ = nested_paths;
```
Comment (alexcrichton) [inline src/cargo/core/workspace.rs:421]: I think it's fine to omit this and just avoid using it here.

File: src/cargo/core/workspace.rs [hunk 5]
```diff
@@ -362,14 +362,107 @@ impl<'cfg> Workspace<'cfg> {
...
+                            name,
+                            /* pkg_id */ None,
+                            source,
+                            &mut nested_paths,
+                            self.config,
+                            &mut warnings,
+                            /* platform */ None,
+                            // NOTE: Since we use ConfigRelativePath, this root isn't used as
+                            // any relative paths are resolved before they'd be joined with root.
+                            self.root(),
```
Comment (alexcrichton) [inline src/cargo/core/workspace.rs:406]: Instead of using `self.root` could this perhaps use something like `/nonexistent` or similar to be a sort of debug-assert that it's not used/

**Review by alexcrichton — COMMENTED** [review]
Apologies for the delay, but this version looks good to me! Ia've got some minor thoughts inline, but otherwise r=me

**Comment by jonhoo** [comment]
@alexcrichton Excellent, all good comments — I've pushed fixes in https://github.com/rust-lang/cargo/pull/9204/commits/b50cde88b458774f0f1f8211ef450ff7872f3894. I don't think bors picked up your r= though?

**Comment by alexcrichton** [comment]
@bors: r+

Oh that's not an actual feature of bors, it's more of "I won't take XX days to review this next time, I'll write r+ soon after I see the email it was updated"

**Comment by ehuss** [comment]
@jonhoo Can you follow up with a PR to add this to the unstable documentation ([here](https://github.com/rust-lang/cargo/blob/master/src/doc/src/reference/unstable.md)).  Unstable features should always be documented there.

**Comment by jonhoo** [comment]
Will do! Should we leave this open as a tracking issue? What do you envision as the stabilization path for this feature?

**Comment by ehuss** [comment]
I think it's fine to open a new issue, it doesn't matter too much.  The "new issue" button should have a template for a tracking issue.

**Comment by ehuss** [comment]
Oh, and I'm not sure exactly what needs to happen to stabilize.  We don't have a clear process for that, since it doesn't happen too often. 😜  Normally we let it bake on nightly for a while, to let people to try it out and see if there are any issues, and to gain more experience with it.  This also gives some time to consider if it is the right design and something that we actually want to commit to.

**Comment by jonhoo** [comment]
Documentation PR: https://github.com/rust-lang/cargo/pull/9270

Tracking issue: https://github.com/rust-lang/cargo/pull/9269
